client/src/app/shared/components/sorted-tree/sorted-tree.component.ts

```diff
@@ -0,0 +1,630 @@
+        }
+    }
+
+    /**
+     * Function to get all nodes under the given node with level + 1.
+     *
+     * @param node The parent of the searched children.
+     *
+     * @returns An array that contains all the nearest children.
+     */
+    public getChildNodes(node: OSFlatNode): OSFlatNode[] {
+        const nodeIndex = this.osTreeData.indexOf(node);
+        const childNodes: OSFlatNode[] = [];
+
+        if (nodeIndex < this.osTreeData.length - 1) {
+            for (let i = nodeIndex + 1; i < this.osTreeData.length && this.osTreeData[i].level >= node.level + 1; ++i) {
+                if (this.osTreeData[i].level === node.level + 1) {
+                    childNodes.push(this.osTreeData[i]);
+                }
+            }
+        }
+
+        return childNodes;
+    }
+
+    /**
+     * Function to look for all nodes that are under the given node.
+     * This includes not only the nearest children, but also the children of the children.
+     *
+     * @param node The parent of the nodes.
+     *
+     * @returns An array containing all the subnodes, inclusive the children of the children.
+     */
+    private getAllSubNodes(node: OSFlatNode): OSFlatNode[] {
+        return this._getAllSubNodes(node, []);
+    }
+
+    /**
+     *
+     * @param node The parent of the nodes.
+     * @param array The existing array containing all subnodes
+     *
+     * @returns An array containing all subnodes, inclusive the children of the children.
+     */
+    private _getAllSubNodes(node: OSFlatNode, array: OSFlatNode[]): OSFlatNode[] {
+        array.push(node);
+        for (const child of this.getChildNodes(node)) {
+            this._getAllSubNodes(child, array);
+        }
+        return array;
+    }
+
+    /**
+     * Function to check the position in the list that is shown.
+     * This is necessary to identify the calculated position from CdkDragDrop.
+     *
+     * @param node The node whose position at the shown list should be checked.
+     *
+     * @returns The calculated position as number.
+     */
+    private getPositionOnScreen(node: OSFlatNode): number {
+        let currentPosition = this.osTreeData.length;
+        for (let i = this.osTreeData.length - 1; i >= 0; --i) {
+            --currentPosition;
+            const parent = this.getExpandedParentNode(this.osTreeData[i]);
+            if (parent === node) {
+                break;
+            } else {
+                i = parent.position;
+            }
+        }
+        return currentPosition;
+    }
+
+    /**
+     * Function to check if the node should render.
+     *
+     * @param node which is viewed.
+     *
+     * @returns boolean if the node should render. Related to the state of the parent, if expanded or not.
+     */
+    public shouldRender(node: OSFlatNode): boolean {
+        const parent = this.getParentNode(node);
+        return !parent || parent.isExpanded;
+    }
+
+    /**
+     * Function, that handles the click on a node.
```

Comment:
When collapse a node with some (mlti leveled) children, all children collapses. So if you expand this node, you have to manually expand every child. Maybe just toggle the expand-state of the affected node. `shouldRender` must call it's parent to check, if it's expanded recursively. This is a lot of work, so I would suggest to cache this result in a private property.
-> when someone clicks on the arrow, recalculate the private property for all items below the clicked node. This new property is a different one to the `isExpanded`, because this gives the state of the arrow.